crates/iota-node/src/lib.rs

```diff
@@ -833,6 +843,34 @@ impl IotaNode {
         let iota_node_metrics =
             Arc::new(IotaNodeMetrics::new(&registry_service.default_registry()));
 
+        // --- Create shared checkpoint broadcast channel and buffer for gRPC and
+        // checkpoint logic ---
+        let (grpc_checkpoint_summary_tx, grpc_checkpoint_data_tx) = if let Some(grpc_api_address) =
+            config.grpc_api_address
+        {
+            let (summary_tx, _) = tokio::sync::broadcast::channel(CHECKPOINT_BROADCAST_BUFFER_SIZE);
+            let (data_tx, _) = tokio::sync::broadcast::channel(CHECKPOINT_BROADCAST_BUFFER_SIZE);
+            let rocks = RocksDbStore::new(
+                cache_traits.clone(),
+                committee_store.clone(),
+                checkpoint_store.clone(),
+            );
+            let rest_read_store = std::sync::Arc::new(RestReadStore::new(state.clone(), rocks));
+            // Use the shared broadcast channel and buffer for gRPC checkpoint streaming
+            let grpc_service =
+                CheckpointGrpcService::new(rest_read_store, summary_tx.clone(), data_tx.clone());
+            tokio::spawn(async move {
+                info!("Starting gRPC server on {grpc_api_address}");
+                tonic::transport::Server::builder()
+                    .add_service(CheckpointServiceServer::new(grpc_service))
+                    .serve(grpc_api_address)
+                    .await
+                    .expect("gRPC server failed");
+            });
+            (Some(summary_tx), Some(data_tx))
```

Comment:
We should refactor that into it's own function. I'm also not sure if a validator should run these APIs at all. So maybe better check how we build the JSON and REST server in `build_http_server` or `start_graphiql_server` and do something similar.

Reply:
Fixed it at [af6baf4](https://github.com/iotaledger/iota/pull/7511/commits/af6baf4986deb9581b92786371f08f1a188e1679). Thanks!